Обработка состояний экрана и Перевод получения избранных на Flow из DAO

manualcheg opened the pull request:
Обработка состояний экрана

Перевод получения избранных на Flow из DAO

manualcheg committed:
1. Создан интерфейс с состояниями Избранного FavoritesState
2. Во ViewModel отправляются через LiveData все варианты данных из БД и обработка ошибки во время получения
3. Во Фрагменте сделана обработка состояний, присылаемых из ViewModel
4. В файле разметки добавлены id для элементов плейсхолдера для обработки всех состояний данных
5. Попытка отдавать Flow из DAO

manualcheg committed:
1. Сменён формат отдаваемый методом getFavorites(): Flow<List<FavoritesVacancyEntity>>
2. Вверх по цепочке интерфейсов и их реализаций передается Flow до ViewModel, где разбирается и обрабатываются List<Vacancy>
3. В FavoritesStorage производится преобразование типа в листе
4. Suspend оставлен только у метода в UseCase
5. Дополнен DI модуль

JudjinGM commented on app/src/main/java/ru/practicum/android/diploma/favorites/domain/FavoritesState.kt:
```diff
@@ -0,0 +1,9 @@
+
+import ru.practicum.android.diploma.common.domain.model.vacancy_models.Vacancy
+
+interface FavoritesState {
```
sealed interface

manualcheg replied:
ok

JudjinGM commented on app/src/main/java/ru/practicum/android/diploma/favorites/data/dataSource/FavoritesStorage.kt:
```diff
@@ -1,13 +1,14 @@
+import kotlinx.coroutines.flow.Flow
 import ru.practicum.android.diploma.common.domain.model.vacancy_models.Vacancy
 
 interface FavoritesStorage {
```
мб переименовать в FavoritesLocalDataSource?

manualcheg replied:
хорошо

JudjinGM commented on app/src/main/java/ru/practicum/android/diploma/favorites/data/repositoryImpl/FavoritesDBRepositoryImpl.kt:
```diff
@@ -17,9 +16,8 @@ class FavoritesDBRepositoryImpl(
         favoritesStorage.deleteFromFavorites(id)
     }
```
Класс и интерфейс имеет смысл переименовать в FavoriteRepository для того, чтобы не было специфики, что это относится к db, тк db это зависимость datasource

manualcheg replied:
Хорошо

JudjinGM commented on app/src/main/java/ru/practicum/android/diploma/favorites/ui/fragment/FavoritesFragment.kt:
```diff
@@ -21,8 +38,48 @@ class FavoritesFragment: Fragment() {
+
     override fun onDestroy() {
         super.onDestroy()
         _binding = null
```
Binding советуют уничтожать в onDestroyView. Здесь же лучше занулить все адаптеры binding.favouritesRecyclerView.adapter = null, vacanciesAdapter = null

manualcheg replied:
добавил

JudjinGM reviewed (commented):
Ревью

manualcheg committed:
Изменения в связи с ревью:
1.1 Переименование FavoritesStorage в FavoritesLocalDataSource
1.2 Переименование FavoritesDBRepositoryImpl в FavoritesRepositoryImpl
2. Интерфейс с состояниями сделан запечатаным (sealed)
3. Сделано обнуление адаптера RecyclerView при уничтожении фрагмента

manualcheg committed:
1. Промежуточное состояние 2

manualcheg committed:
1. Отмена Промежуточное состояние 2

manualcheg committed:
Merge branch 'dev' into feature/FAVR-108/create_processing_state_of_screen
